tensorflow_graphics/projects/pointnet/train.py

```diff
@@ -0,0 +1,148 @@
+#Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Training loop for PointNet v1 on modelnet40."""
+# pylint: disable=missing-function-docstring
+
+import tensorflow as tf
+from tqdm import tqdm
+import helpers
+import augment
+from tensorflow_graphics.datasets.modelnet40 import ModelNet40
+from tensorflow_graphics.nn.layer.pointnet import PointNetVanillaClassifier as PointNet
+
+# ------------------------------------------------------------------------------
+# ------------------------------------------------------------------------------
+# ------------------------------------------------------------------------------
+
+parser = helpers.ArgumentParser()
+parser.add("--batch_size", 32)
+parser.add("--num_epochs", 250)
+parser.add("--num_points", 2048, help="subsampled (max 2048)")
+parser.add("--learning_rate", 1e-3, help="initial Adam learning rate")
+parser.add("--lr_decay", True, help="use Adam learning rate decay")
+parser.add("--bn_decay", .5, help="batch norm decay momentum")
+parser.add("--tb_every", 100, help="tensorboard frequency (itereations)")
+parser.add("--ev_every", 308, help="evaluation frequency (itereations)")
+parser.add("--tfds", True, help="use TFDS dataset loader")
+parser.add("--augment", True, help="use augmentations")
+parser.add("--tqdm", True, help="enable the progress bar")
+FLAGS = parser.parse_args()
+
+# ------------------------------------------------------------------------------
+# ------------------------------------------------------------------------------
+# ------------------------------------------------------------------------------
+
+ds_train, info = ModelNet40.load(split="train", with_info=True)
+num_examples = info.splits["train"].num_examples
+ds_train = ds_train.shuffle(num_examples, reshuffle_each_iteration=True)
+ds_train = ds_train.repeat(FLAGS.num_epochs)
+ds_train = ds_train.batch(FLAGS.batch_size)
+num_classes = info.features["label"].num_classes
```

Comment:
prefetch? It's not the most complex pipeline, but I'm yet to find a good reason not to put it as the last step.

Reply:
I am going to do this in two stages, starting from something that works, and then do a second PR for performance. I also wanted to move the augmentation pipeline from the training loop down to a "map" function, and he prefetch will make more sense then.